### ISSUE TITLE
Support string and number filter operators in dashboard parameter filters 

### ISSUE THREAD

**Issue by daltojohnso** [metabase/metabase#15055]
As a front-end engineer, I need an endpoint to support additional filter operators for dashboard parameter filters.

The filter operators needed*, per UI mocks, are
- string
  - `=`
  - `!=`
  - `contains`
  - `does-not-contain`
  - `starts-with`
  - `ends-with`
- number
  - `=`
  - `!=`
  - `between`
  - `>=`
  - `<=`

The `string` operators are needed by filters with the types `category` or `location/*` (where `*` is `city`, `state`, `zip_code`, or `country`). The `number` operators are needed by a yet-to-be added filter option, "Number."

Currently, parameter filters only support the equivalent of an `=` operator. When a user adds a value to a parameter filter in the UI, the client sends a POST request to the `/api/card/:card-id/query` endpoint with a body containing the parameter filter's `type`, `target`, and `value`:

```
{
  "parameters": [
    {
      "type":"category",
      "target":["dimension",["field",1024,null] ],
      "value":["Foo","Bar"]
    }
  ]
}
```

A few questions:
1. I've been unable to glean in what ways the parameter `type` (ie `category` or `location/city`) affects the behavior of the filter beyond, perhaps, changing which values appear for it in dropdowns. Does the `type` have any effect on the generated query?
2. Does the front-end require any other additional API changes to support the adding of operators to parameters? The current way we save parameters is the client POSTs (or PUTs) to the dashboard endpoint a `parameters` list with objects like:
```
{
    id: "5e0b4726"
    name: "Category"
    slug: "category"
    type: "category"
}
```
It seems like the dashboard save/update endpoints accept `parameters` with properties beyond the ones listed here, I just want to be sure these objects are open for extension.

\* since this list is a subset of potential question filter operators, perhaps it makes sense to also expose any other potential operators since some are missing from current mocks (for example, `!=` is missing as a potential string filter operator)

**Comment by paulrosenzweig** [comment]
>  Does the type have any effect on the generated query?

The type is important for date filters. e.g. `date/relative` filters let you pass `past30days`. I'm not aware of how type would affect the query for any non-date filters.

---
Another endpoint update:
In addition the card query endpoint, the chained filter endpoints probably also need to be aware of the new parameter types/options. If a dashboard has two linked filters A -> B. Currently, the search/dropdown for B will be filtered based on A. The filtering on A should probably match the behavior (contains, ends-with, etc) that A filters dashcards with.

**Comment by camsaul** [comment]
you want a new endpoint for this?

**Comment by paulrosenzweig** [comment]
I think it should extend the existing chain filter and query endpoints.

**Comment by daltojohnso** [comment]
@camsaul It seems like we can use the same `card/card-id/query` endpoint but with changes to the body of the POST request since all we do right now is pass a `value` string or string array

**Comment by camsaul** [comment]
We already have a `:number`, and a `:text` parameter type

https://github.com/metabase/metabase/blob/master/src/metabase/driver/common/parameters/values.clj#L28-L34

and

https://github.com/metabase/metabase/blob/caecc7bb8d2ec8aaee87159d01f39556d3cb5913/frontend/src/metabase/query_builder/components/template_tags/TagEditorHelp.jsx#L72-L79

for example

**Comment by daltojohnso** [comment]
Oh nice did not see that ID was a number type

**Comment by camsaul** [comment]
It sounds like there are two things we need to add here:

1.  Some way to specify the new widgets types when you define the parameter. e.g. "this parameter should get a widget that lets you pick the filter operator" or "this parameter should get a widget that is locked to 'starts-with'"
2.  Some way to pass the selected operator in when you run the query (in the `:parameters` map

We already have a precedent for 1. When you create a temporal filter widget, you can choose from different types of `date` widget such as `date/all-options` or `date/month-year` (see https://github.com/metabase/metabase/blob/28ce196d5a33aa134f41813d5a6d235ca89d9ca6/frontend/src/metabase/parameters/components/ParameterValueWidget.jsx#L32-L39 for example). This widget `:type` is ignored on the backend IIRC

For number 2, temporal filters pass in different string representations of things, e.g. `past2weeks` or `2021-03~2021-04` and we can deduce the correct type of filter clause to generate based on the format of the string. Remember that filter values can be passed either as part of `:value` in the `:parameters` key of a query itself, or in the query parameters when viewing a dashboard, e.g. `?my_param=past2weeks`. So currently all filter values need a way to be encoded as strings in a human-friendly way, so we can deduce the filter type from the strings themselves.

**Comment by camsaul** [comment]
It would probably be easiest if you just change the FE code to pass in the parameter type used when it passes in `:parameters`, and leave `:value` as-is, e.g.

Perhaps adding another key to `:parameters` to specify the value that should be used e.g.

```js
{
    id: "5e0b4726",
    target: ["template-tag", "abc"],
    value: "BBQ",
    type: "text/starts-with" ; as opposed to "text" or whatever
}
```

For query params maybe you could use two keys like

```
?abc=BBQ&abc.type=text%2Fstarts-with
```

I think that would do the trick. Maybe you can combine those into one key somehow... you'll have to think about that.

Either way, once we have a way to get the new types we can generate appropriate filter clauses on the backend for them

**Comment by camsaul** [comment]
Parameter value encoding/decoding code would be a good candidate for the shared backend/frontend CLJS lib as opposed to duplicating the code between the two codebases. Both backend and frontend have to understand the parameters so we need that code in both places...

maybe concentrate on part 1 (new widgets types) since that's pretty much all frontend anyway to start with and that will give us a bit of time to figure out how to make part 2 part of the shared lib

**Comment by daltojohnso** [comment]
OK, will do, thanks a lot for the help, @camsaul!

**Comment by daltojohnso** [comment]
@camsaul -- Do the parameter types `location/(city|state|country|zip-code)` and `category` mean anything to the back-end or are they only used on the front-end to filter the fields available for mapping? Seems like I should split the parameter `type` into two separate props so that we can avoid a bunch of `location/starts-with`, `category/starts-with` etc. and replace all of these with `text/starts-with` etc.

**Comment by daltojohnso** [comment]
I've gone the route of adding new types like `"text/stars-with"` and `"number/between"`. The query params seem to handle every case but the `all-options` case already, so I will need to think about how to encode which value is selected.

**Comment by daltojohnso** [comment]
LMK if this makes more sense on the backend to be `text/*` instead of `category/*` and `location/*` -- from the FE perspective it was easier to implement that since these field parameters were already sending types of `category` and `location/city` etc.

**Comment by daltojohnso** [comment]
For the new `all-options` parameter types (`number/all-options`, `category/all-options`, `location/all-options`), my most likely approach is to treat the selected operator as an additional value in the first entry of the `values` array. The FE can shift it off the beginning of the array to build out the correct type before sending it to the backend.

An alternative to that is sending `category/all-options` as a parameter type to the BE with a value array like `['starts-with', 'foo']` or for `number/all-options` `['between', 1, 111]`. Would that make sense?

**Comment by camsaul** [comment]
@daltojohnso the array would totally work. How would you do it in a URL? URL-encode it? The only downside is that wouldn't be as human-readable

For date ranges we currently do `start~end`, e.g. `2021-03-01~2021-04-01`. I think that could work for number ranges too, but I don't know how you could make special syntax work for string starts-with vs ends-with for example. I guess you'd have to URL-encode the strings anyway....

I think ideally we'd try to minimize the syntax differences between date filters and the new string/number ones. Or bring the old date stuff in line with the new filters at some point (maybe as a follow-on refactor). Better to have one way we do stuff than two

**Comment by dpsutton** [comment]
One question, we have a clear boundary that the endpoint can translate whatever shape the parameters come in to a more normalized for that the query processor can handle. That way the qp doesn't need to worry about url encoding or treating the first value in the values arg as an operator (and we avoid the ambiguity thing, whether `"BETWEEN"` is a value to match on or an operator.

@daltojohnso if this is the case, can you give me your preferred shape to send the api to indicate that a number column is between two values and a preferred shape for a string column starts with a substring?

**Comment by dpsutton** [comment]
If my understanding is correct, we just update `metabase.query-processor.middleware.parameters.mbql/build-filter-clause` to look for the operation that is passed down and where the args for operations go. For design considerations, will these be additive? For instance, click on a few values in the UI for equality check and then add string begins with "foo"? 

ie, 
```clojure
{:type "category"
 :target ["dimension"
          ["field"
           26
           {:source-field 5}]]
 :value ["Doohickey" "Bob"]} ;; where category name is doohickey or bob
{:filter [:or
          [:= [:field 26 {:source-field 5}] "Doohickey"]
          [:= [:field 26 {:source-field 5}] "Bob"]]}
```

would one of the options in `:value` be starts with "foo" to get all foobars?

**Comment by daltojohnso** [comment]
I think what works best for the FE is

for a string `starts-with` filter
```
{:type "category/starts-with"
 :target ["dimension"
          ["field"
           26
           {:source-field 5}]]
 :value ["foo"]} 
```

for a number `between` filter
```
{:type "number/between"
 :target ["dimension"
          ["field"
           26
           {:source-field 5}]]
 :value [1, 100]} 
```

Defining new parameter `type` values plays well with the code we currently have and we get query param -> widget input population for free with the exception of the new `all-options` widgets. As for `all-options`, my plan is to treat the selected option like a value on the front-end so that it gets set as a query param like so:

`?all_options_1=between&all_options_1=1&all_options_1=100`
`?all_options_2=starts-with&all_options_2=foo`

This will let me populate operators like `starts-with` and `ends-with` correctly. I'll add some code to replace `number/all-options` and `['between', 1, 100]` with `number/between` and `[1, 100]` before sending it off.

**Comment by daltojohnso** [comment]
I think the above approach removes a lot of the complexity you were asking about. In the UI the flow is:

```
1. click to add a new parameter widget
2. first screen, select either a "Number" or a "Category" or a "Location" option
3. second screen, select the type of operator -- "Starts with" or "Between" or "All options"
4. map the parameter widget to some fields, save dashboard
```
This will create a new parameter with a type of `number/between` or `category/starts-with` or `location/=` (I'm just using the names of operators already defined for field filters).

**Comment by camsaul** [comment]
@dpsutton I think you just need to add some new code here that handles the new param types (text/number/whatever) https://github.com/metabase/metabase/blob/master/src/metabase/query_processor/middleware/parameters/mbql.clj#L48-L52

You might want to put the code for converting the new types to MBQL clauses in their own namespaces. This is the date param handling code https://github.com/metabase/metabase/blob/master/src/metabase/driver/common/parameters/dates.clj. So maybe adding `string` and `number` sibling namespaces would make sense.  

I don't know if you'll need to do anything else besides that... I'd start with that stuff and see what happens. If it's not working yet ping me and we'll figure out what else we need to tweak

**Comment by dpsutton** [comment]
@camsaul thanks. yeah i'll see if it makes sense to make three namespaces or if its better as just one. if they're really small i'll let the future break them up. Do you know how native interacts with this? I see there's a namespace for parameters for native queries but didn't know if that's unrelated and more for snippet expansion.

**Comment by daltojohnso** [comment]
Resolved by https://github.com/metabase/metabase/pull/15068

**Comment by sahil-malhotra** [comment]
Hi, I just upgraded Metabase to 0.39.0.1 but can't see the string and number filter operators in the "Add filters" menu.<img width="730" alt="Screenshot" src="https://user-images.githubusercontent.com/16478068/115529898-eda6af80-a2b0-11eb-90d5-8089bba88a9e.png">

Is it available only in the enterprise release?

**Comment by flamber** [comment]
@sahil-malhotra The functionality has been hidden, since it was not completed, but will be part of version 0.40.0

**Comment by maximepvrt** [comment]
do you have more visibility on the version that will integrate this functionality ? I don't see that in the v41 RC1

**Comment by flamber** [comment]
@maximepvrt It's currently planned for 0.42

**Comment by rlotun** [comment]
this has been addressed and will be in 0.42
